openapi-diff/src/modeler/AutoRest.Swagger/Model/Operation.cs

```diff
@@ -243,46 +219,172 @@ private void CheckParameters(ComparisonContext<ServiceDefinition> context, Opera
-                if (newParam == null) continue;
-
-                
                 SwaggerParameter oldParam = FindParameterEx(newParam, priorOperation.Parameters, previousRoot.Parameters);
 
-                if (oldParam == null)
+                if (oldParam != null)
+                    continue;
+
+                // Did not find required parameter in the old swagger i.e. required parameter is added
+                context.PushItemByName(newParam.Name);
+
+                context.LogBreakingChange(
+                    newParam.IsRequired
+                        ? ComparisonMessages.AddingRequiredParameter
+                        : ComparisonMessages.AddingOptionalParameter,
+                    newParam.Name);
+
+                context.Pop();
+            }
+            context.Pop();
+        }
+
+        private static (SwaggerParameter param, bool isGlobal) ResolveParam(SwaggerParameter param, ServiceDefinition serviceDef)
+            => string.IsNullOrWhiteSpace(param.Reference)
+                ? (param, false)
+                : (FindReferencedParameter(param.Reference, serviceDef.Parameters), true);
+
+        /// <summary>
+        /// This method reports breaking change on parameters whose order matters and it has changed.
+        ///
+        /// This method works by first filtering out all parameters whose order does not matter.
+        /// To see how do we determine if parameter order matters, consult the ParamOrderMatters method.
+        ///
+        /// Once we have the collection of current and prior params whose order matters, we compare their order
+        /// by index. If at any point current and prior params differ (see implementation of this method
+        /// to see how we determine if two params differ), then we report a breaking change.
+        ///
+        /// If a new parameter, whose order matters, was added to current version, we will not report any
+        /// breaking changes on it as it didn't exist before. However, its addition may still shift
+        /// other parameters' order, and that will be reported as breaking change. E.g.:
+        /// [Foo, Bar] -> [Qux, Foo, Bar].
+        /// Here Qux has been added thus shifting Foo and Bar's order.
+        /// 
+        /// If a parameter definition has changed from prior version in such way that its order now matters
+        /// but previously didn't, then we assume its order matters. E.g. the parameter was converted from
+        /// "client" to "method".
+        ///
+        /// If a parameter definition has changed from prior version in such way that its order previously
+        /// mattered, but now it does not, then we assume order does not matter. E.g. the parameter
+        /// was converted from "method" to "client".
+        /// 
+        /// Additional context provided at:
+        /// https://github.com/Azure/azure-sdk-tools/issues/7170#issuecomment-2162156876
+        /// </summary>
+        private void DetectChangedParameterOrder(
```

Comment:
@mikekistler please have a look at this method - at least at the comment.

Reply:
Regarding a parameter changing from order mattering to not, or vice-versa, I think this should always be flagged as a breaking change.  I also think you should determine whether a parameter's order matters for each version independently.

Reply:
@mikekistler

Re:

> Regarding a parameter changing from order mattering to not, or vice-versa, I think this should always be flagged as a breaking change.

I guess you would like for it to be [a new rule](https://github.com/Azure/openapi-diff/tree/main/docs#readme), yes? Or you are OK with reporting this in case the "parameter order matter-ness" changes?

``` csharp
        public static MessageTemplate ChangedParameterOrder = new MessageTemplate
        {
            Id = 1042,
            Code = nameof(ComparisonMessages.ChangedParameterOrder),
            Message = "The order of parameter '{0}' was changed. ",
            Type = MessageType.Update
        };
```

I was thinking to create a rule similar to this one:

[1025 - RequiredStatusChange](https://github.com/Azure/openapi-diff/blob/main/docs/rules/1025.md) but name it `ParameterOrderRequirementChange`.


Re this:

> I also think you should determine whether a parameter's order matters for each version independently.

I am not sure what you mean by that. I think the code a I implemented it in the method `DetectChangedParameterOrder` does this already.